tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs

```diff
@@ -0,0 +1,159 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class WasmLoadAssembliesAndReferences : Task
+{
+    public string? MainAssembly { get; set; }
+
+    // If true, continue when a referenced assembly cannot be found.
+    // If false, throw an exception.
+    public bool SkipMissingAssemblies { get; set; }
+
+    // Either one of these two need to be set
+    public string[]? AssemblySearchPaths { get; set; }
+    public string[]? Assemblies { get; set; }
+    public string[]? ExtraAssemblies { get; set; }
```

Comment:
Actually, `MainAssembly` has no special significance here. This task is essentially taking bunch of assemblies as input, finding all the references to those, and returning the full set. And for resolving the references it has two cases:
1. specified AssemblySearchPaths
2. or, find `*Private.CoreLib.dll`, and look in it's containing directory

So, essentially, you can make this:
```CSharp
     // If true, continue when a referenced assembly cannot be found.
     // If false, throw an exception.
     public bool SkipMissingAssemblies { get; set; }

    // Either one of these two need to be set
    public string[]? AssemblySearchPaths { get; set; }

    [Required]
    [NotNull]
    public string[]? Assemblies { get; set; }
```

And in the Execute method, instantiate the `resolver`, depending on whether you have `AssemblySearchPaths` or not. And rest of the code to iterate over the assemblies should be the same. Something like:

```csharp
        foreach (var asm in Assemblies)
        {
            var asmFullPath = Path.GetFullPath(asm);
            try
            {
                var refAssembly = mlc.LoadFromAssemblyPath(asmFullPath);
                Add(mlc, refAssembly);
            }
            catch (Exception ex) when (ex is FileLoadException || ex is BadImageFormatException || ex is FileNotFoundException)
            {
                if (SkipMissingAssemblies)
                    Log.LogMessage(MessageImportance.Low, $"Loading extra assembly '{asm}' failed with {ex}. Skipping");
                else
                {
                    Log.LogError($"Failed to load assembly from ExtraAssemblies '{asm}': {ex}");
                    return false;
                }
            }
        }

```

Reply:
And a follow up to that: msbuild already has a task that does this, it's called `ResolveAssemblyReference`, which can used here as:
```xml
    <ResolveAssemblyReference
      AssemblyFiles="bin\WasmSample.dll;$(MicrosoftNetCoreAppRuntimePackDir)lib\$(NetCoreAppCurrent)\System.Private.Runtime.InteropServices.JavaScript.dll"
      SearchPaths="@(AssemblySearchPaths)">
      <Output TaskParameter="ResolvedFiles" ItemName="ReferencedAssemblies" />
      <Output TaskParameter="ResolvedDependencyFiles" ItemName="ReferencedAssemblies" />
    </ResolveAssemblyReference>
```

.. instead of a new task. You could try this out. Check that you get the same list of assemblies as from *Searcher task.  If you find anything missing, then build with `/v:diag`, and that will show you why something didn't get resolved. https://docs.microsoft.com/en-us/visualstudio/msbuild/resolveassemblyreference-task?view=vs-2019

Reply:
It does seem like there is redundant code that serve to iterate over a set of assemblies.
From what I understand, the two paths for the original task are:
1. If provided `AssemblySearchPaths`, ensure that all paths are valid, and then add the main assembly and its referenced assemblies. (From what I hear, `ExtraAssemblies` is a way to overcome the scenario where the linker removed essential assemblies which will be fixed when the real linker is used with `<PublishTrimmed>true</Publishtrimmed>`)
2. Without `AssemblySearchPaths`, its expected that `Assemblies` contains all essential assemblies to build the wasm app (of which, one assembly is expected to be `System.Private.CoreLib.dll`.

By refactoring the two cases into a standard process of 
```Figure out the right `_resolver` -> Iterate over a set of assemblies```
Then, if the `MainAssembly` and `ExtraAssemblies` parameters are removed now should need to be folded into `Assemblies` when `AssemblySearchPaths` != `null` outside the scope of the task.
Then is the expectation of the task to be of two forms
```
<WasmLoadAssembliesAndReferences
   ...
   AssemblySearchPaths="@(AssemblySearchPaths)"
   Assemblies="$(MainAssembly);@(ExtraAssemblies)" />
```
```
<WasmLoadAssembliesAndReferences
   ...
   <!-- No AssemblySearchPaths -->
   Assemblies="@(Assemblies)" />
```

Reply:
Yeah, that's what I had in mind.

Reply:
Looks like `ResolveAssemblyReference` is not loading some referenced assemblies. 
For `debugger-test.dll` which depends on `System.Console`, the resolved primary and reference assemblies only contains `debugger-test.dll` and `debugger-test.pdb`
For `WasmTestRunner.dll` which depends on atleast `System.Runtime`, the resolved reference assemblies are
```
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/Microsoft.DotNet.XHarness.TestRunners.Xunit.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Collections.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.CoreLib.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/Microsoft.DotNet.XHarness.TestRunners.Common.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/xunit.abstractions.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/xunit.runner.utility.netcoreapp10.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/xunit.execution.dotnet.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/xunit.core.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.Xml.Linq.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.Xml.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Linq.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Console.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.IO.FileSystem.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/Mono.Options.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Net.Sockets.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Collections.Concurrent.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Linq.Expressions.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.ObjectModel.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Diagnostics.TraceSource.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Text.RegularExpressions.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Net.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Net.Http.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Collections.NonGeneric.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Collections.Specialized.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Runtime.CompilerServices.Unsafe.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.Uri.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.Runtime.InteropServices.JavaScript.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Runtime.InteropServices.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.ComponentModel.TypeConverter.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Security.Cryptography.Algorithms.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/Microsoft.Win32.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Diagnostics.DiagnosticSource.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.ComponentModel.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Drawing.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.ComponentModel.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Runtime.Serialization.Formatters.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Security.Cryptography.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Runtime.Numerics.dll
```
which doesn't contain `System.Runtime.dll`

Reply:
The problem here is mismatched versions. The sample is referencing 5.0 assemblies, but we are trying to resolve them against 6.0 ones. `AutoUnify=true` helps with that, partially. I was working on a fix locally. I will try to clean that up, and share.

Reply:
btw, you can run with `/v:diag`, and check the messages from `ResolveAssemblyReference` task, to see details of why it didn't resolve `System.Console`, which should show the version issue.